src/psbt/finalizer.rs

```diff
@@ -0,0 +1,369 @@
+/// script is correctly interpreted according to the context
+/// The psbt must have included final script sig and final witness.
+/// In other words, this checks whether the finalized psbt interprets
+/// correctly
+pub fn interpreter_check(psbt: &Psbt) -> Result<(), Error> {
+    for (index, input) in psbt.inputs.iter().enumerate() {
+        let script_pubkey =
+            get_scriptpubkey(psbt, index).map_err(|e| Error::InputError(e, index))?;
+        let empty_script_sig = Script::new();
+        let empty_witness = Vec::new();
+        let script_sig = input.final_script_sig.as_ref().unwrap_or(&empty_script_sig);
+        let witness = input
+            .final_script_witness
+            .as_ref()
+            .unwrap_or(&empty_witness);
+        // get the descriptor and stack(unification of scriptsig and witness)
+        let (des, stack) = from_txin_with_witness_stack(script_pubkey, &script_sig, &witness)
+            .map_err(|e| Error::InputError(InputError::MiniscriptError(e), index))?;
+
+        // Now look at all the satisfied constraints. If everything is filled in
+        // corrected, there should be no errors
+        let cltv = psbt.global.unsigned_tx.lock_time;
+        let csv = psbt.global.unsigned_tx.input[index].sequence;
+        // TODO: After sighash is merged by #116, change this to verify the secp signature
+        // over the sighash
+        let sighash_ty = input.sighash_type.unwrap_or(SigHashType::All);
+        let sighash = match des {
+            Descriptor::Wsh(..)
+            | Descriptor::Wpkh(..)
+            | Descriptor::ShWsh(..)
+            | Descriptor::ShWpkh(..) => {
+                // None defaults to All
+                let amt = get_amt(psbt, index).map_err(|e| Error::InputError(e, index))?;
+                let mut sighash_cache = SigHashCache::new(&psbt.global.unsigned_tx);
+                sighash_cache.signature_hash(index, &des.witness_script(), amt, sighash_ty)
+            }
+            _ => psbt.global.unsigned_tx.signature_hash(
+                index,
+                &des.witness_script(),
+                sighash_ty.as_u32(),
+            ),
+        };
+        let secp = secp256k1::Secp256k1::verification_only();
+
+        let msg = secp256k1::Message::from_slice(&sighash[..])
+            .map_err(|e| Error::InputError(InputError::MiniscriptError(MsError::from(e)), index))?;
+
+        // we have already checked the sighash flags in sanity check step
+        let vfyfn = |pk: &bitcoin::PublicKey, (sig, _)| secp.verify(&msg, &sig, &pk.key).is_ok();
+        let constraints_iter = SatisfiedConstraints::from_descriptor(&des, stack, vfyfn, cltv, csv);
+        let constraints: Result<Vec<_>, _> = constraints_iter.collect();
+        constraints
+            .map_err(|e| Error::InputError(InputError::MiniscriptError(MsError::from(e)), index))?;
+    }
+    Ok(())
+}
+
+pub fn finalize(psbt: &mut Psbt) -> Result<(), super::Error> {
```

Comment:
Mutably borrowing the PSBT will leave it in a potentially invalid state in case of an error. I'd rather consume and return the PSBT.

Reply:
We do check the validity of the psbt before returning. 

Somehow, it seems to be that allocating a new psbt when we are only adding two fields per input(final sciptsig and final witness) seems wrong.

Reply:
While working on PSBTs in rust-bitcoin, I am not getting why finalize shouldn't be part of `PartiallySignedTransaction` structure with access to `self`
https://github.com/rust-bitcoin/rust-bitcoin/pull/472/files#diff-77e0d8cfeb8d4f84f6f901f6fc14139fR337

Reply:
I think the method finalize could belong there, but rust-miniscript offers a way to construct the final witness and final scriptsig in a generic way by the satisfier trait. 

We don't want to introduce a rust-miniscript dependancy to rust-bitcoin. 

Maybe one way might be to have the method at both places, the generic finalizer in rust-miniscript and one method in rust-bitcoin that only fills in the required fields.

Reply:
The more I think about this, I think there definitely should be a finalize method in rust-bitcoin that works for all scripts. Miniscript is more of helper to the finalizer. The job of a finalizer is to fill in the final values for witness and script_sig even if they are not miniscripts. 

I think the API should be as follows:

rust-miniscript:
```
fn finalize_miniscript(&mut psbt) -> (Script, Vec<u8>) 
// Final script_sig and witness
// Only works when all scripts used are miniscripts
```

rust-bitcoin:
```
fn finalize(&mut self, witness: Vec<u8>, script_sig: Script)
// This fills in the values in psbt.
```

Reply:
> `fn finalize(&mut self, witness: Vec<u8>, script_sig: Script)`

Do I understand it correctly that this finalize method would work on per-input basis? Otherwise the witness/script sig had to be matched to an input which would make the finalization at least `O(n^2)`. This also means that miniscript `finalize` had to return multiple `(Script, Vec<u8>)` tuples (one for each input).

Reply:
Yup, my bad. Edited the original post. We can discuss the API even further. Maybe it should only work on a per input basis, but I wanted to share my opinion on how to the split between rust-miniscript's role and rust-bitcoin's role on finalizing.